File: kotlin/observer/WeatherRama/observer/displays/StatisticsDisplay.kt
package observer.displays

import observer.Display
import observer.Observer
import subject.WeatherStation

class StatisticsDisplay(weatherStation: WeatherStation) : Observer, Display {

    private var maxTemperature = 0.0f
    private var minTemperature = 200f
    private var temperatureSum = 0.0f
    private var numReadings = 0

    private var weatherStation = WeatherStation()

    init {
        this.weatherStation = weatherStation
        weatherStation.registerObserver(this)
    }

    override fun update(temperature: Float, humidity: Float, pressure: Float) {

        temperatureSum += temperature
        numReadings++

        if (temperature > maxTemperature) { maxTemperature = temperature }

        if (temperature < minTemperature) { minTemperature = temperature }

        display()
    }

    override fun display() {
        println("StatisticsDisplay:\n"
                + "Avg temperature = " + (temperatureSum/numReadings) + "\n"
                + "Max temperature = " + maxTemperature + "\n"
                + "Min temperature = " + minTemperature + "\n"
        )
    }

}

File: python/observer/WeatherRama/observer/displays/StatisticsDisplay.py
from observer.Observer import Observer
from observer.Display import Display
from subject.WeatherStation import WeatherStation


class StatisticsDisplay(Observer, Display):

    def __init__(self, weather_station):
        self.__temperature_sum = 0
        self.__num_readings = 0
        self.__max_temperature = 0
        self.__min_temperature = 200
        self.__weather_station = WeatherStation()
        self.__weather_station = weather_station
        self.__weather_station.register_observer(self)

    def update(self, temperature, humidity, pressure):
        self.__temperature_sum += temperature
        self.__num_readings += 1

        if temperature > self.__max_temperature: self.__max_temperature = temperature
        if temperature < self.__min_temperature: self.__min_temperature = temperature

        self.display()

    def display(self):
        avg = self.__temperature_sum/self.__num_readings
        print("StatisticsDisplay:\n"
              + "Avg temperature = " + str(avg) + "\n"
              + "Max temperature = " + str(self.__max_temperature) + "\n"
              + "Min temperature = " + str(self.__min_temperature) + "\n")
